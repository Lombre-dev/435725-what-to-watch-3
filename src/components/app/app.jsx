import PropTypes from 'prop-types';
import React from 'react';
import {BrowserRouter, Route, Switch} from 'react-router-dom';
import Main from '../main/main';
import MovieDetailInfo from '../movie-detail-info/movie-detail-info';
import {Movie} from '../types';

export default class App extends React.PureComponent {

  constructor(props) {
    super(props);

    this.state = {
      detailInfoMovie: undefined,
    };

    this._handleMovieListItemClick = this._handleMovieListItemClick.bind(this);
  }

  _handleMovieListItemClick({movie}) {
    this.setState(() => {
      return {
        detailInfoMovie: movie,
      };
    });
  }

  _renderState() {

    const {detailInfoMovie} = this.state;
    const {currentMovie, movies} = this.props;

    if (detailInfoMovie) {
      return (
        <MovieDetailInfo
          movie={detailInfoMovie}
        />
      );
    }

    return (
      <Main
        currentMovie={currentMovie}
        movies={movies}
        onMovieListItemClick={this._handleMovieListItemClick}
      />
    );
  }

  render() {

    const {movies} = this.props;

    return (
      <BrowserRouter>
        <Switch>
          <Route exact path="/">{
            this._renderState()
          }</Route>
          <Route exact path="/dev-movie-detail-info">
            <MovieDetailInfo
              movie={movies[0]}
            />
          </Route>
        </Switch>
      </BrowserRouter>
    );
  }
}

App.propTypes = {
  currentMovie: Movie.isRequired,
  movies: PropTypes.arrayOf(Movie),
};
