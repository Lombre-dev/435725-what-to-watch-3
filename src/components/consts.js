export const GENRES = [
  `Comedy`,
  `Crime`,
  `Documentary`,
  `Drama`,
  `Horror`,
  `Kids & Family`,
  `Romance`,
  `Sci-Fi`,
  `Thriller`,
];

export const RATING_SCORE_LIMIT = 10;
export const RATING_LEVELS = [
  `Bad`,
  `Normal`,
  `Good`,
  `Very good`,
  `Awesome`,
];
