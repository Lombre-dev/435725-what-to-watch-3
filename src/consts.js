export const ALL_GENRE = `All genres`;
const COMEDY_GENRE = `Comedy`;
const CRIME_GENRE = `Crime`;
const DOCUMENTARY_GENRE = `Documentary`;
const DRAMA_GENRE = `Drama`;
const HORROR_GENRE = `Horror`;
const KIDS_AND_FAMILY_GENRE = `Kids & Family`;
const ROMANCE_GENRE = `Romance`;
const SCI_FI_GENRE = `Sci-Fi`;
const THRILLER_GENRE = `Thriller`;

export const AppPages = {
  MAIN: `/`,
  LOGIN: `/login`,
  MY_LIST: `/mylist`,
  MOVIES: `/films`,
  PLAYER: `/player`,
};

export const GENRES = [
  COMEDY_GENRE,
  CRIME_GENRE,
  DOCUMENTARY_GENRE,
  DRAMA_GENRE,
  HORROR_GENRE,
  KIDS_AND_FAMILY_GENRE,
  ROMANCE_GENRE,
  SCI_FI_GENRE,
  THRILLER_GENRE,
];

export const GenreLabels = {
  [COMEDY_GENRE]: `Comedies`,
  [DRAMA_GENRE]: `Dramas`,
  [THRILLER_GENRE]: `Thrillers`,
};

export const MORE_LIKE_THIS_MOVIES_LIMIT = 4;
export const CATALOG_MOVIES_PER_PAGE_LIMIT = 8;

export const SMALL_MOVIE_CARD_PREVIEW_DELAY = 1000;

export const MOVIE_INFO_TABS = [
  `Overview`,
  `Details`,
  `Reviews`,
];

export const PlayerState = {
  INITED: `inited`,
  LOADING: `loading`,
  PLAYING: `playing`,
  PAUSED: `paused`,
  ENDED: `ended`,
};

export const SERVER_URL = `https://htmlacademy-react-3.appspot.com/wtw`;
export const SERVER_RESPONSE_TIMEOUT = 5000;
export const SERVER_USE_COOKIES = true;
export const ServerErrors = {
  NOT_FOUND: 404,
  UNAUTHORIZED: 401,
};
