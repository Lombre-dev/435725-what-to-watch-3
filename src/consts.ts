export const ALL_GENRE = `All genres`;
const COMEDY_GENRE = `Comedy`;
const CRIME_GENRE = `Crime`;
const DOCUMENTARY_GENRE = `Documentary`;
const DRAMA_GENRE = `Drama`;
const HORROR_GENRE = `Horror`;
const KIDS_AND_FAMILY_GENRE = `Kids & Family`;
const ROMANCE_GENRE = `Romance`;
const SCI_FI_GENRE = `Sci-Fi`;
const THRILLER_GENRE = `Thriller`;

export enum AppPages {
  MAIN = `/`,
  LOGIN = `/login`,
  MY_LIST = `/mylist`,
  MOVIES = `/films`,
  PLAYER = `/player`,
}

export const GENRES = [
  COMEDY_GENRE,
  CRIME_GENRE,
  DOCUMENTARY_GENRE,
  DRAMA_GENRE,
  HORROR_GENRE,
  KIDS_AND_FAMILY_GENRE,
  ROMANCE_GENRE,
  SCI_FI_GENRE,
  THRILLER_GENRE,
];
export const GENRES_CATALOG_LIMIT = 9;

export const GenreLabels = {
  [COMEDY_GENRE]: `Comedies`,
  [DRAMA_GENRE]: `Dramas`,
  [THRILLER_GENRE]: `Thrillers`,
};

export const MORE_LIKE_THIS_MOVIES_LIMIT = 4;
export const CATALOG_MOVIES_PER_PAGE_LIMIT = 8;

export const SMALL_MOVIE_CARD_PREVIEW_DELAY = 1000;

export const MOVIE_INFO_TABS = [
  `Overview`,
  `Details`,
  `Reviews`,
];

export enum PlayerState {
  INITED = `inited`,
  LOADING = `loading`,
  PLAYING = `playing`,
  PAUSED = `paused`,
  ENDED = `ended`,
}

export const ServerConfig = {
  URL: `https://htmlacademy-react-3.appspot.com/wtw`,
  RESPONSE_TIMEOUT: 5000,
  USE_COOKIES: true,
};
export enum ServerErrors {
  NOT_FOUND = 404,
  BAD_REQUEST = 400,
  UNAUTHORIZED = 401,
}

export enum AuthorizationErrorCode {
  INCORRECT_LOGIN = 1,
  AUTHORIZATION_FAIL = 2,
}

export const AuthorizationErrorMessage = {
  [AuthorizationErrorCode.INCORRECT_LOGIN]: `Please enter a valid email address`,
  [AuthorizationErrorCode.AUTHORIZATION_FAIL]: `We canâ€™t recognize this email and password combination. Please try again.`,
};

export enum AuthorizationStatus {
  NO_AUTH = `unauthorized`,
  AUTH = `authorized`,
}

export const REVIEW_RATING_STARS_LIMIT = 5;
export const REVIEW_COMMENT_MIN_LENGTH = 50;
export const REVIEW_COMMENT_MAX_LENGTH = 400;

export enum LoadingDataStatus {
  LOADING = `loading`,
  READY = `ready`,
  ERROR = `error`,
}
