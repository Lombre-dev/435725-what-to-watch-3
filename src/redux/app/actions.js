import {createAction} from 'redux-act';

export const setAppLoadingStart = createAction(`SET_APP_LOADING_START`);
export const setAppMovies = createAction(`SET_APP_MOVIES`);
export const setAppLoadingError = createAction(`SET_APP_LOADING_COMPLETE`);
