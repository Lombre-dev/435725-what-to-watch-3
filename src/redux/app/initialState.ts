export const initialState = {
  status: undefined,
  movies: [],
};
