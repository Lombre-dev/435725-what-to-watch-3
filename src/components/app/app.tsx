import * as React from 'react';
import {connect} from 'react-redux';
import {BrowserRouter, Route, Switch} from 'react-router-dom';
import {AppPages, LoadingDataStatus} from '../../consts';
import {Operations} from '../../redux/app/operations';
import {getAppStatus} from '../../redux/app/selectors';
import AddReviewPage from '../add-review-page';
import CatalogPage from '../catalog-page/catalog-page';
import LoadingDataBlock from '../loading-data-block/loading-data-block';
import MoviePage from '../movie-page/movie-page';
import MyListPage from '../my-list-page/my-list-page';
import NotFoundPage from '../not-found-page/not-found-page';
import PlayerPage from '../player-page';
import PrivateRoute from '../private-route/private-route';
import ProtectedRoute from '../protected-route/protected-route';
import SignInPage from '../sign-in-page/sign-in-page';

type TAppProps = {
  status?: LoadingDataStatus;
  onMount?: Function;
};

class App extends React.PureComponent<TAppProps> {

  public componentDidMount() {

    const {onMount} = this.props;

    onMount();
  }

  public render() {

    const {status} = this.props;

    if (status !== LoadingDataStatus.READY) {
      return <LoadingDataBlock status={status} />;
    }

    return (
      <BrowserRouter>
        <Switch>
          <ProtectedRoute exact path={AppPages.MAIN} render={() => <CatalogPage />} />
          <Route exact path={AppPages.LOGIN} component={SignInPage} />
          <PrivateRoute exact path={AppPages.MY_LIST} render={() => <MyListPage />} />
          <PrivateRoute exact path={`${AppPages.MOVIES}/:id/review`} render={() => <AddReviewPage />} />
          <ProtectedRoute exact path={`${AppPages.MOVIES}/:id`} render={() => <MoviePage />} />
          <Route exact path={`${AppPages.PLAYER}/:id`} component={PlayerPage} />
          <Route component={NotFoundPage} />
        </Switch>
      </BrowserRouter >
    );
  }
}

function mapStateToProps(state: object) {
  return {
    status: getAppStatus(state),
  };
}

function mapDispatchToProps(dispatch: Function) {
  return {
    onMount: () => {
      dispatch(Operations.init());
    },
  };
}

export {App};
export default connect(mapStateToProps, mapDispatchToProps)(App);
