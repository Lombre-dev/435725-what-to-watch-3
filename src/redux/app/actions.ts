import {createAction} from 'redux-act';
import {TMovie} from '../../components/types';

export const setAppLoadingStart = createAction(`SET_APP_LOADING_START`);
export const setAppMovies: (movies: TMovie[]) => void = createAction(`SET_APP_MOVIES`);
export const setAppLoadingError = createAction(`SET_APP_LOADING_COMPLETE`);
