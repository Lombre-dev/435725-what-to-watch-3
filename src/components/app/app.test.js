import React from 'react';
import {Provider} from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import {AuthorizationStatus, LoadingDataStatus} from '../../consts';
import {App} from './app';

const GENRES = [`Drama`, `Comedy`, `Kids & Family`];
const CURRENT_GENRE = GENRES[0];
const MOVIES = [
  {
    id: 0,
    title: `The Grand Budapest Hotel`,
    genres: [`Drama`],
    year: 2014,
    poster: `img/the-grand-budapest-hotel-poster.jpg`,
    frames: [`img/the-grand-budapest-hotel-poster.jpg`],
    ratingScore: 8.9,
    ratingReviewsCount: 240,
    preview: `./samples/sintel_trailer-480p.mp4`,
    src: `./samples/sintel_trailer-480p.mp4`,
    description: `In the 1930s, the Grand Budapest Hotel is a popular European ski resort, presided over by concierge Gustave H. (Ralph Fiennes). Zero, a junior lobby boy, becomes Gustave's friend and protege.`,
    story: `Gustave prides himself on providing first-class service to the hotel's guests, including satisfying the sexual needs of the many elderly women who stay there. When one of Gustave's lovers dies mysteriously, Gustave finds himself the recipient of a priceless painting and the chief suspect in her murder.`,
    director: `Wes Andreson`,
    duration: 99,
    actors: [
      `Bill Murray`,
      `Edward Norton`,
      `Jude Law`,
      `Willem Dafoe`,
      `Some Actor 1`,
      `Some Actor 2`,
    ],
  },
  {
    id: 1,
    title: `Bohemian Rhapsody`,
    genres: [`Drama`],
    year: 2014,
    poster: `img/bohemian-rhapsody.jpg`,
    frames: [`img/bohemian-rhapsody.jpg`],
    duration: 120,
    preview: ``,
    src: ``,
    ratingScore: 7.5,
    ratingReviewsCount: 100,
    description: `Description`,
    story: `Story`,
    director: `Director`,
    actors: [
      `Some Actor 1`,
      `Some Actor 2`,
      `Some Actor 3`,
      `Some Actor 4`,
      `Some Actor 5`,
    ],
  },
  {
    id: 2,
    title: `Macbeth`,
    genres: [`Drama`],
    year: 2014,
    preview: ``,
    src: ``,
    poster: `img/macbeth.jpg`,
    frames: [`img/macbeth.jpg`],
    duration: 100,
    ratingScore: 6.9,
    ratingReviewsCount: 50,
    description: `Description`,
    story: `Story`,
    director: `Director`,
    actors: [
      `Some Actor 1`,
      `Some Actor 2`,
      `Some Actor 3`,
      `Some Actor 4`,
      `Some Actor 5`,
    ],
  },
];
const HAS_MORE_MOVIES = true;
const LOADING_DATA_STATUS = LoadingDataStatus.LOADING;
const HANDLE_EVENT = () => {};

const mockStore = configureStore([thunk]);

describe(`<App />`, () => {

  it(`render should be match markup`, () => {

    const store = mockStore({
      app: {
        status: LOADING_DATA_STATUS,
      },
      movie: {
        movie: undefined,
        moviesLike: [],
      },
      catalog: {
        promoMovie: MOVIES[0],
        genre: CURRENT_GENRE,
        genres: GENRES,
        movies: MOVIES,
        hasMoreMovies: HAS_MORE_MOVIES,
      },
      user: {
        id: -1,
        name: undefined,
        email: undefined,
        avatar: undefined,
        authError: undefined,
        authStatus: AuthorizationStatus.NO_AUTH,
      }
    });

    const result = renderer
      .create(<Provider store={store}>
        <App
          status={LOADING_DATA_STATUS}
          onMount={HANDLE_EVENT}
        />
      </Provider>, {
        createNodeMock: () => {
          return {};
        }
      })
      .toJSON();

    expect(result).toMatchSnapshot();
  });
});
