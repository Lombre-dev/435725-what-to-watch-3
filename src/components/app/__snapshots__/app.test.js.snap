// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> render should be match markup 1`] = `
<div
  style={
    Object {
      "color": "#dfcf77",
      "fontSize": "30px",
      "fontWeight": "bold",
      "height": "100%",
      "width": "100%",
    }
  }
>
  loading
</div>
`;
